bundle/src/main/java/com/adobe/acs/commons/rewriter/impl/TemplatedTransformer.java

```diff
@@ -0,0 +1,146 @@
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.util.List;
+import java.util.Map;
+
+import com.adobe.acs.commons.properties.PropertyAggregatorService;
+import com.adobe.acs.commons.properties.util.TemplateReplacementUtil;
+import com.adobe.acs.commons.rewriter.ContentHandlerBasedTransformer;
+import com.adobe.granite.rest.Constants;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.rewriter.ProcessingComponentConfiguration;
+import org.apache.sling.rewriter.ProcessingContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.AttributesImpl;
+
+public class TemplatedTransformer extends ContentHandlerBasedTransformer {
+
+    private static final Logger log = LoggerFactory.getLogger(TemplatedTransformer.class);
+
+    private Map<String, Object> properties;
+    private PropertyAggregatorService aggregatorService;
+
+    public TemplatedTransformer() {
+    }
+
+    public TemplatedTransformer(PropertyAggregatorService propertyAggregatorService) {
+        this.aggregatorService = propertyAggregatorService;
+    }
+
+    @Override
+    public void init(ProcessingContext processingContext, ProcessingComponentConfiguration processingComponentConfiguration) throws IOException {
+        SlingHttpServletRequest request = processingContext.getRequest();
+        PageManager pageManager = request.getResourceResolver().adaptTo(PageManager.class);
+
+        // Get the combined properties via service
+        Page page = pageManager.getContainingPage(request.getResource());
+        if (page != null) {
+            properties = aggregatorService.getProperties(page);
+        }
+    }
+
+    public void startElement(String uri, String localName, String quaName, Attributes atts) throws SAXException {
+        if (shouldRun() && localName.equals("a")) {
+            AttributesImpl newAttrs = new AttributesImpl(atts);
+            for (int i = 0; i < newAttrs.getLength(); i++) {
+                String currentAttribute = decode(newAttrs.getValue(i));
+                if (TemplateReplacementUtil.hasPlaceholder(currentAttribute)) {
+
+                    // Get the current placeholder in the string
+                    String placeholder = TemplateReplacementUtil.getPlaceholder(currentAttribute);
+
+                    // Transform it to the key in the property map
+                    String key = TemplateReplacementUtil.getKey(placeholder);
+
+                    // If the placeholder key is in the map then replace it
+                    if (properties.containsKey(key)) {
+                        String replaceValue = (String) properties.get(key);
+                        newAttrs.setValue(i, currentAttribute.replace(placeholder, encode(replaceValue)));
```

Comment:
Should this instead be `newAttrs.setValue(i, encode(currentAttribute.replace(placeholder, replaceValue))` since the entire attribute was decoded and thus should be re-encoded?

Reply:
very good point - updated

Reply:
Were you able to fully test this?

To test if this is automatically decoding I would put a value that includes a " into an attribute and see if this class sees " or instead an encoded value.

Then I would try using a replacement value that includes a " and see if the resulting HTML encodes the " or if the HTMl ends up broken.

Reply:
Yeah I tested it with invalid characters, but it doesn't seem like there's a good fix-all solution here. 
If I encode the whole property I get broken authored paths when the authored value contains any path at all: `/content/page.html?source={{mySite.source}}` to `%2Fcontent%2Fpage.html%3Fsource%3Dword%20space`
or
`{{mySite.internalPath}}` to `%2Fcontent%2Fpage.html`
or even
`{{mySite.externalPath}}` to `https%3A%2F%2Fwww.google.com%2F`
If I encode only the value I'm replacing then it only supports query params and not paths (which was the original intention of the encoding):
`/content/page.html?source={{mySite.source}}` to `/content/page.html?source=word%20space`
I could add some sort of universal setting, but I think this could be solved with a feature like the `dateFormat` discussed above - something like `{{mySite.internalPath [encode=true]}}` (or something). Right now it's in the hands of authors to not break stuff (which is not ideal), but without having encoding be optional I'm not sure how it can be resolved. 
I could encode the HTML entities so that things like `"` would turn into `&quot;` (or `&#x22;`) if that would be beneficial, but I think it would still result in broken links (better than broken markup though?)

Reply:
I wonder how HTL escapes HTML attributes correctly?  If we could use that code, that might be ideal.

Barring that, I think we could do something like escape ', ", & (to prevent accidental HMTL entities), and < > (in case its an attribute that someone didnt put quotes around).  We would want to escape these only in the replacement values, leaving the rest of the string as-is.

If a `href` variable is "https://www.google.com?search=abc&quot;xyz" does the link when clicked change the `&quot;` to the correct encoded value?